Sync optimizations

schaal opened the pull request:
This is not ready to be merged yet, I wanted to get some feedback first for those changes before I clean it up.

These are some experiments to improve the time needed for synchronizations:

* Use [OkHttp](http://square.github.io/okhttp/) instead of HTTPUrlConnection
  * Also use OkHttp to download images
  * This also simplifies the network code a bit
* Start AsyncTasks in parallel
  * Synchronize the completion of all tasks using a [CountDownLatch](https://developer.android.com/reference/java/util/concurrent/CountDownLatch.html)
* Only get the api version on the first sync

A quick benchmark shows an improvement on average from 4,68 s to 3,20 s needed for synchronization:
```
origin/master
-------------
V/OwnCloudSyncService﹕ Time needed (sync) 00:00:04.906
V/OwnCloudSyncService﹕ Time needed (sync) 00:00:04.624
V/OwnCloudSyncService﹕ Time needed (sync) 00:00:04.505
V/OwnCloudSyncService﹕ Time needed (sync) 00:00:04.703
V/OwnCloudSyncService﹕ Time needed (sync) 00:00:04.679

Average: 4,68 s

sync-optimizations
----------
V/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.288
V/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.305
V/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.128
V/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.134
V/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.156

Average: 3,20 s
```

David-Development commented:
Removing IReader interface: Great, this was a fragment from the first implementation of multiple api endpoints.. Not needed any longer. Thanks

One suggestion to probably make the sync even faster:
Encapsulate every execute call in OwnCloud_Reader AsyncTask by AsyncTaskHelper.StartAsyncTask(new [AsyncTask...]);
This could increase the performance as well. Because they can be executed in parallel.

What are the advantages of using the okhttp lib? (haven't used it before)

schaal commented:
I feel it has a nicer API and it has supposedly better performance. It also supports connection pooling, which reduces the latency when connecting multiple times to the same host.

I'll try using the AsyncTaskHelper and check whether it improves performance further.

schaal commented:
I now cleaned up the patch set for you to review. I didn't include the parallelization of the AsyncTasks as it doesn't seem to improve things much and makes it more fragile.

Updated benchmark:
```
origin/master @ 7c646e3e
-------------
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:04.540
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:04.480
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:04.489
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:04.420
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:04.420

Average: 4,47 s

sync-optimizations
----------
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.249
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.100
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.099
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.100
de.luhmer.owncloudnewsreader D/OwnCloudSyncService﹕ Time needed (sync): 00:00:03.090

Average: 3,13 s
```

David-Development commented:
Great, Thank you!
